Multiple SSH Keys support

davidjumani opened the pull request:
### Description

Is a continuation of https://github.com/apache/cloudstack/pull/5215
The result of the GSoC project 
Adds support for multiple SSH Keys for VMs.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

### Screenshots (if appropriate):

![Screenshot from 2022-02-08 16-18-45](https://user-images.githubusercontent.com/8244774/152972425-bc0c545f-3131-4437-b541-a9038b900c79.png)

![Screenshot from 2022-02-08 16-18-25](https://user-images.githubusercontent.com/8244774/152972438-013a14e4-af67-42da-9f57-98872ea121c0.png)

![Screenshot from 2022-02-10 13-01-23](https://user-images.githubusercontent.com/8244774/153358829-9c378072-24c5-4aa0-8cd9-ba22ca34965f.png)

davidjumani commented:
@blueorangutan package

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 2562

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_multiplication_x: debian :heavy_check_mark: suse15. SL-JID 2564

nvazquez commented:
@blueorangutan package

blueorangutan commented:
@nvazquez a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 2568

davidjumani commented:
@blueorangutan test

blueorangutan commented:
@davidjumani a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-3285)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 30891 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5965-t3285-kvm-centos7.zip
Smoke tests completed. 91 look OK, 1 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_invalid_upgrade_kubernetes_cluster | `Failure` | 28.31 | test_kubernetes_clusters.py
test_02_upgrade_kubernetes_cluster | `Failure` | 26.92 | test_kubernetes_clusters.py
test_03_deploy_and_scale_kubernetes_cluster | `Failure` | 25.89 | test_kubernetes_clusters.py
test_04_autoscale_kubernetes_cluster | `Failure` | 27.86 | test_kubernetes_clusters.py
test_05_basic_lifecycle_kubernetes_cluster | `Failure` | 29.01 | test_kubernetes_clusters.py
test_06_delete_kubernetes_cluster | `Failure` | 35.24 | test_kubernetes_clusters.py
test_07_deploy_kubernetes_ha_cluster | `Failure` | 36.99 | test_kubernetes_clusters.py
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 35.97 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 36.01 | test_kubernetes_clusters.py
ContextSuite context=TestKubernetesCluster>:teardown | `Error` | 131.59 | test_kubernetes_clusters.py

DaanHoogland commented:
> Will run / fix the kubernetes tests after #5863 is merged

does that mean this is/should be still draft @davidjumani ?

davidjumani commented:
@DaanHoogland Nope, just that it's not worth fixing as the CKS codebase is changed in that PR

davidjumani commented:
@blueorangutan package

davidjumani commented:
@blueorangutan package

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 2628

davidjumani commented:
@blueorangutan test

blueorangutan commented:
@davidjumani a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-3363)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 30213 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5965-t3363-kvm-centos7.zip
Smoke tests completed. 91 look OK, 1 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_invalid_upgrade_kubernetes_cluster | `Failure` | 33.25 | test_kubernetes_clusters.py
test_02_upgrade_kubernetes_cluster | `Failure` | 33.01 | test_kubernetes_clusters.py
test_03_deploy_and_scale_kubernetes_cluster | `Failure` | 33.96 | test_kubernetes_clusters.py
test_04_autoscale_kubernetes_cluster | `Failure` | 34.00 | test_kubernetes_clusters.py
test_05_basic_lifecycle_kubernetes_cluster | `Failure` | 31.02 | test_kubernetes_clusters.py
test_06_delete_kubernetes_cluster | `Failure` | 35.19 | test_kubernetes_clusters.py
test_07_deploy_kubernetes_ha_cluster | `Failure` | 34.20 | test_kubernetes_clusters.py
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 28.82 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 27.84 | test_kubernetes_clusters.py
ContextSuite context=TestKubernetesCluster>:teardown | `Error` | 124.23 | test_kubernetes_clusters.py

nvazquez commented:
@blueorangutan package

blueorangutan commented:
@nvazquez a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 2648

DaanHoogland commented:
@blueorangutan test

blueorangutan commented:
@DaanHoogland a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

nvazquez commented:
Ping for review @Pearl1594 @weizhouapache

blueorangutan commented:
<b>Trillian test result (tid-3387)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 41827 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5965-t3387-kvm-centos7.zip
Smoke tests completed. 91 look OK, 1 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_invalid_upgrade_kubernetes_cluster | `Failure` | 26.94 | test_kubernetes_clusters.py
test_02_upgrade_kubernetes_cluster | `Failure` | 28.02 | test_kubernetes_clusters.py
test_03_deploy_and_scale_kubernetes_cluster | `Failure` | 24.82 | test_kubernetes_clusters.py
test_04_autoscale_kubernetes_cluster | `Failure` | 29.78 | test_kubernetes_clusters.py
test_05_basic_lifecycle_kubernetes_cluster | `Failure` | 26.75 | test_kubernetes_clusters.py
test_06_delete_kubernetes_cluster | `Failure` | 35.11 | test_kubernetes_clusters.py
test_07_deploy_kubernetes_ha_cluster | `Failure` | 35.15 | test_kubernetes_clusters.py
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 27.86 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 35.01 | test_kubernetes_clusters.py
ContextSuite context=TestKubernetesCluster>:teardown | `Error` | 131.68 | test_kubernetes_clusters.py

davidjumani commented:
@blueorangutan package

davidjumani commented:
@blueorangutan package

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 2666

davidjumani commented:
@blueorangutan test

blueorangutan commented:
@davidjumani a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

weizhouapache commented:
@davidjumani 
code looks good to me. tested ok.

I have few questions or suggestions

(1) for existing vms which has `SSH.PublicKey` in user_vm_details before cloudstack upgrade, would be good to add a record `SSH.KeyPairNames` in the table ?

(2) the URL of ssh keypair page is http://localhost:5050/#/ssh/<ssh keypair name>. What if there are multiple ssh keypairs with same name on the platform ? uuid looks better than ssh keypair name.

(3) on UI, can the existing ssh keypair be checked by default when open the dialog of reset ssh keypair ?

weizhouapache commented:
@nvazquez @davidjumani 
we can merge this PR and address the comments (https://github.com/apache/cloudstack/pull/5965#issuecomment-1046626256) afterwards. it is fine with me.

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
@weizhouapache a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 2684

weizhouapache commented:
@blueorangutan test

blueorangutan commented:
@weizhouapache a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

davidjumani commented:
I'll address the first point in this PR since we would like to maintain backward compatibility. I can raise another for the other ones

blueorangutan commented:
<b>Trillian test result (tid-3410)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 30205 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5965-t3410-kvm-centos7.zip
Smoke tests completed. 91 look OK, 1 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_invalid_upgrade_kubernetes_cluster | `Failure` | 27.11 | test_kubernetes_clusters.py
test_02_upgrade_kubernetes_cluster | `Failure` | 25.86 | test_kubernetes_clusters.py
test_03_deploy_and_scale_kubernetes_cluster | `Failure` | 26.79 | test_kubernetes_clusters.py
test_04_autoscale_kubernetes_cluster | `Failure` | 27.84 | test_kubernetes_clusters.py
test_05_basic_lifecycle_kubernetes_cluster | `Failure` | 38.98 | test_kubernetes_clusters.py
test_06_delete_kubernetes_cluster | `Failure` | 34.88 | test_kubernetes_clusters.py
test_07_deploy_kubernetes_ha_cluster | `Failure` | 27.77 | test_kubernetes_clusters.py
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 28.84 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 28.90 | test_kubernetes_clusters.py
ContextSuite context=TestKubernetesCluster>:teardown | `Error` | 122.99 | test_kubernetes_clusters.py

nvazquez commented:
@davidjumani seems like the failures are consistent, can you check if it needs fixing?

davidjumani commented:
@blueorangutan package

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

davidjumani commented:
@blueorangutan package

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

davidjumani commented:
10.0.35.53

davidjumani commented:
@blueorangutan package

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

weizhouapache commented on engine/schema/src/main/resources/META-INF/db/schema-41610to41700.sql:
```diff
@@ -641,3 +641,8 @@ CREATE VIEW `cloud`.`domain_router_view` AS
             and async_job.instance_type = 'DomainRouter'
             and async_job.job_status = 0;
+
+INSERT INTO `cloud`.`user_vm_details`(`vm_id`, `name`, `value`)
```
@davidjumani 
is it possible that ssh keypairs are under different account but have same public key ?
for example, a user register a public key under ROOT/admin, then register same public key in his owner account ?

davidjumani replied:
It is possible but an edge case since users would not register the same key twice. I don' think we need to consider this scenario

weizhouapache replied:
@davidjumani 
I think it would be better to consider all possible scenarios.

my sugguestion is

```
INSERT INTO `cloud`.`user_vm_details`(`vm_id`, `name`, `value`)
SELECT `user_vm_details`.`vm_id`, 'SSH.KeyPairNames', `ssh_keypairs`.`keypair_name`
        FROM `cloud`.`user_vm_details`
        INNER JOIN `cloud`.`ssh_keypairs` ON ssh_keypairs.public_key = user_vm_details.value
        INNER JOIN `cloud`.`vm_instance` ON vm_instance.id = user_vm_details.vm_id
        WHERE ssh_keypairs.account_id = vm_instance.account_id
```

weizhouapache replied:
@davidjumani @nvazquez 
what's your opinion on this comment ?

davidjumani replied:
@weizhouapache Makes sense. I'll make the changes

andrijapanicsb replied:
Was this addressed? This is NOT an edge case, I have customers using this setup extensively.

There is also a unique key constraint in the table, so pls make sure this is not a problem

davidjumani replied:
Yes it's been addressed

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_check_mark: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 2697

davidjumani commented:
@blueorangutan package

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

davidjumani commented:
@blueorangutan ui

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: http://qa.cloudstack.cloud:8080/client/pr/5965 (SL-JID-1181)

davidjumani commented:
> @davidjumani code looks good to me. tested ok.
> 
> I have few questions or suggestions
> 
> (1) for existing vms which has `SSH.PublicKey` in user_vm_details before cloudstack upgrade, would be good to add a record `SSH.KeyPairNames` in the table ?
> 
> (2) the URL of ssh keypair page is http://localhost:5050/#/ssh/. What if there are multiple ssh keypairs with same name on the platform ? uuid looks better than ssh keypair name.
> 
> (3) on UI, can the existing ssh keypair be checked by default when open the dialog of reset ssh keypair ?

I've addressed points 1 & 3 in this PR. Point 2 might require a larger change since the uuid field was recently added to ssh keyparis

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 2698

davidjumani commented:
5151

davidjumani commented:
@blueorangutan test

blueorangutan commented:
@davidjumani a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-3420)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 36647 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5965-t3420-kvm-centos7.zip
Smoke tests completed. 92 look OK, 0 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

davidjumani commented:
@blueorangutan package

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 2706

davidjumani commented:
@blueorangutan test

blueorangutan commented:
@davidjumani a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

weizhouapache reviewed (approved):
thanks @davidjumani 

I have tested this again, everything looks good.
cc @nvazquez

blueorangutan commented:
<b>Trillian test result (tid-3428)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 45704 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5965-t3428-kvm-centos7.zip
Smoke tests completed. 91 look OK, 1 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_07_deploy_kubernetes_ha_cluster | `Failure` | 3622.52 | test_kubernetes_clusters.py
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 0.07 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 0.04 | test_kubernetes_clusters.py
ContextSuite context=TestKubernetesCluster>:teardown | `Error` | 69.29 | test_kubernetes_clusters.py

weizhouapache commented:
> Trillian test result (tid-3428) Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7 Total time taken: 45704 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5965-t3428-kvm-centos7.zip Smoke tests completed. 91 look OK, 1 have errors Only failed tests results shown below:
> 
> Test	Result	Time (s)	Test File
> test_07_deploy_kubernetes_ha_cluster	`Failure`	3622.52	test_kubernetes_clusters.py
> test_08_upgrade_kubernetes_ha_cluster	`Failure`	0.07	test_kubernetes_clusters.py
> test_09_delete_kubernetes_ha_cluster	`Failure`	0.04	test_kubernetes_clusters.py
> ContextSuite context=TestKubernetesCluster>:teardown	`Error`	69.29	test_kubernetes_clusters.py

@davidjumani 
this needs more investigation.

might it be caused by systemvm template /cloud-init ?

andrijapanicsb commented:
LGTM, tested briefly and it looks good.

davidjumani commented:
@weizhouapache Seems unrelated. I'll kick the tests again

davidjumani commented:
@blueorangutan test

blueorangutan commented:
<b>Trillian test result (tid-3444)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 35994 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5965-t3444-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_clusters.py

davidjumani commented:
@blueorangutan package

blueorangutan commented:
@davidjumani a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 2727

davidjumani commented:
@blueorangutan test

blueorangutan commented:
@davidjumani a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-3446)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 31716 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5965-t3446-kvm-centos7.zip
Smoke tests completed. 92 look OK, 0 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
